x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportProfileHasPrivilegesAction.java

```diff
@@ -0,0 +1,131 @@
+        final AuthorizationEngine.PrivilegesToCheck privilegesToCheck = new AuthorizationEngine.PrivilegesToCheck(
+            Arrays.asList(request.clusterPrivileges()),
+            Arrays.asList(request.indexPrivileges()),
+            Arrays.asList(request.applicationPrivileges())
+        );
+        resolveApplicationPrivileges(request, ActionListener.wrap(applicationPrivilegeDescriptors -> {
+            profileService.getProfileSubjects(Arrays.asList(request.profileUids()), ActionListener.wrap(profileSubjectsAndFailures -> {
+                threadPool.generic().execute(() -> {
+                    final List<String> hasPrivilegeProfiles = Collections.synchronizedList(new ArrayList<>());
+                    final List<String> errorProfiles = Collections.synchronizedList(new ArrayList<>(profileSubjectsAndFailures.v2()));
+                    final Runnable allDone = () -> listener.onResponse(
+                        new ProfileHasPrivilegesResponse(hasPrivilegeProfiles.toArray(new String[0]), errorProfiles.toArray(new String[0]))
+                    );
+                    final Collection<Map.Entry<String, Subject>> profileUidAndSubjects = profileSubjectsAndFailures.v1().entrySet();
+                    final AtomicInteger counter = new AtomicInteger(profileUidAndSubjects.size());
+                    if (counter.get() == 0) {
+                        allDone.run();
+                        return;
+                    }
+                    for (Map.Entry<String, Subject> profileUidToSubject : profileUidAndSubjects) {
+                        final String profileUid = profileUidToSubject.getKey();
+                        final Subject subject = profileUidToSubject.getValue();
+                        authorizationService.checkPrivileges(
+                            subject,
+                            privilegesToCheck,
+                            applicationPrivilegeDescriptors,
+                            ActionListener.wrap(privilegesCheckResult -> {
+                                if (privilegesCheckResult.allMatch()) {
+                                    hasPrivilegeProfiles.add(profileUid);
+                                }
+                                if (counter.decrementAndGet() == 0) {
+                                    allDone.run();
+                                }
+                            }, checkPrivilegesException -> {
+                                logger.debug(
+                                    new ParameterizedMessage("Failed to check privileges for profile [{}]", profileUid),
+                                    checkPrivilegesException
+                                );
+                                errorProfiles.add(profileUid);
+                                if (counter.decrementAndGet() == 0) {
+                                    allDone.run();
+                                }
+                            })
+                        );
```

Comment:
I am surprised that we don't have an existing ActionListener subtype that does this aggregation and handling. Maybe worthwhile to extract the pattern here?

Reply:
Honestly I haven't even looked for one. I know about the `GroupedActionListener` but that one is more strict about the erros. I think the abstraction we would need here would have to be a base class for `GroupedActionListener`.

Although I like the idea, I feel like I've expended the inovation budged with this PR, and would propose to defer it for another time.